### PR TITLE
add projections simple test

### PR DESCRIPTION
**Pull request by AKS1996**
in view of #24. missing methods for extracting matrices A,b,c

### CODE REVIEW

File: src/MOI_wrapper.jl [hunk 1]
```diff
@@ -129,6 +130,26 @@ function MOI.set(model::Optimizer, attr::MOI.ConstraintFunction, ci::CI{F, S}, f
...
+function MOI.set(model::Optimizer, attr::MOI.ConstraintSet, ci::CI{F, S}, s::S) where {F<:SUPPORTED_VECTOR_FUNCTIONS, S<:SUPPORTED_VECTOR_SETS}
+    return MOI.set(model.optimizer,attr,ci,s)
+end
+
+function MOI.get(model::Optimizer, attr::MOI.ConstraintFunction, ci::CI{F, S}) where {F<:SUPPORTED_VECTOR_FUNCTIONS, S<:SUPPORTED_VECTOR_SETS}
+    return MOI.get(model.optimizer, attr, ci)
+end
+
+function MOI.set(model::Optimizer, attr::MOI.ConstraintFunction, ci::CI{F, S}, f::F) where {F<:SUPPORTED_VECTOR_FUNCTIONS, S<:SUPPORTED_VECTOR_SETS}
+    return MOI.set(model.optimizer,attr,ci,f)
```
Comment (matbesancon) [inline src/MOI_wrapper.jl:150]:
```suggestion
    return MOI.set(model.optimizer, attr, ci, f)
```

File: src/MOI_wrapper.jl [hunk 2]
```diff
@@ -418,3 +443,63 @@ end
 function MOI.modify(model::Optimizer, ci::CI{F, S}, chg::MOI.AbstractFunctionModification) where {F<:MOI.VectorAffineFunction{Float64}, S <: SUPPORTED_VECTOR_SETS}
     MOI.modify(model.optimizer, ci, chg)
 end
+
+"""
+    Method to differentiate optimal solution `x`, `y`, `s`
+"""
+function backward_conic!(model::Optimizer)
+    if MOI.get(model, MOI.TerminationStatus()) in [MOI.LOCALLY_SOLVED, MOI.OPTIMAL]
+        @assert MOI.get(model.optimizer, MOI.SolverName()) == "SCS"
```
Comment (matbesancon) [inline src/MOI_wrapper.jl:452]: is this verification necessary?

File: src/utils.jl [hunk 3]
```diff
@@ -133,3 +133,69 @@ function get_problem_data(model::MOI.AbstractOptimizer)
...
+            result[1] = 1.0
+            result[2:length(z)] = x / norm_x
+            result *= (norm_x + t) / 2.0
+            return result
+        end
+    end
+end
+
+"""
+    derivative of the projection of vector `z` on MOI set `cone`
```
Comment (matbesancon) [inline src/utils.jl:165]: You should specify how is the derivative expressed, is it a matrix with D[i,j] = d proj_i / d  z_i

File: src/utils.jl [hunk 4]
```diff
@@ -133,3 +133,69 @@ function get_problem_data(model::MOI.AbstractOptimizer)
     
     return Q, q, G, h, A, b, nz, var_idx, nineq, ineq_con_idx, neq, eq_con_idx
 end
+
+
+"""
+    projection of vector `z` on MOI set `cone`
+"""
+function π(cone::MOI.AbstractVectorSet, z::Array{Float64})
+    if cone isa MOI.Zeros
```
Comment (joaquimg) [inline src/utils.jl:142]: Dispatch on cone types instead of doing the conditionals

File: test/MOI_wrapper.jl [hunk 5]
```diff
@@ -532,3 +532,30 @@ end
...
+    MOI.optimize!(model)
+
+    grad_projection = backward_conic!(model)
+
+    @test grad_projection ≈ [1.          0.          0.          0.          0.        ;
+                   0.          1.          0.          0.          0.        ;
+                   0.          0.          0.5         0.35355328 -0.3535535 ;
+                   0.          0.          0.35355328  0.54289328  0.04289326;
+                   0.          0.         -0.3535535   0.04289326  0.54289323] atol=ATOL rtol=RTOL
+end
```
Comment (joaquimg) [inline test/MOI_wrapper.jl:561]: you got these results from python?

File: src/utils.jl [hunk 6]
```diff
@@ -133,3 +133,69 @@ function get_problem_data(model::MOI.AbstractOptimizer)
...
+        return zeros(Float64, size(z))
+    elseif cone isa MOI.Nonnegatives
```
Comment (joaquimg) [inline src/utils.jl:144]: you can also add Nonpositives, the maths should be very simple

File: test/MOI_wrapper.jl (hunk 5 above)
Comment (AKS1996) [reply test/MOI_wrapper.jl:561]: yes. can add a simple `print` command in https://github.com/cvxgrp/diffcp

File: src/utils.jl (hunk 6 above)
Comment (AKS1996) [reply src/utils.jl:144]: ok. will do it

File: src/MOI_wrapper.jl (hunk 2 above)
Comment (AKS1996) [reply src/MOI_wrapper.jl:452]: actually not :sweat_smile: i didn't knew if any other solver also expresses the matrices A,b,c as SCS.jl does

File: src/utils.jl [hunk 7]
```diff
@@ -133,3 +133,69 @@ function get_problem_data(model::MOI.AbstractOptimizer)
...
+            result *= (norm_x + t) / 2.0
+            return result
+        end
+    end
+end
+
+"""
+    derivative of the projection of vector `z` on MOI set `cone`
+"""
+function Dπ(cone::MOI.AbstractVectorSet, z::Array{Float64})
```
Comment (matbesancon) [inline src/utils.jl:167]: The function is type-unstable, with sometimes a matrix, sometimes a vector being returned

Comment (blegat) [reply src/utils.jl:167]: Is it an issue if it's only the union of two types ?

File: src/utils.jl (hunk 3 above)
Comment (AKS1996) [reply src/utils.jl:165]: added

File: src/utils.jl (hunk 7 above)
Comment (AKS1996) [reply src/utils.jl:167]: actually for `BlockDiagonals` to work, i need to turn vectors returned by this method into matrices

Comment (AKS1996) [reply src/utils.jl:167]: will change it here only. that way the function will be type stable too

File: test/MOI_wrapper.jl (hunk 5 above)
Comment (joaquimg) [reply test/MOI_wrapper.jl:561]: add that as a comment here

File: src/utils.jl (hunk 4 above)
Comment (joaquimg) [reply src/utils.jl:142]: for instance:
```julia
function π(cone::MOI.Zeros, z::Array{Float64})
...
end
function π(cone::MOI.Nonnegatives, z::Array{Float64})
...
end
```

Comment (AKS1996) [reply src/utils.jl]: done

**Comment by AKS1996** [comment]
don't know why travis is not showing up. but [build works fine](https://travis-ci.org/github/AKS1996/DiffOpt.jl)

File: test/MOI_wrapper.jl (hunk 5 above)
Comment (joaquimg) [reply test/MOI_wrapper.jl:571]: @AKS1996 just add this and goto go for me!

Comment (AKS1996) [reply test/MOI_wrapper.jl:573]: added

**Pull request merged**